text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+Like the Peer, it performs standard validation logic when a new block of transactions arrives, but it also creates and stores a cryptographic hash of each new key-value pair of the blockchain state. This makes it possible to verify that the data coming from the Peer is correct -- remember that in the FPC setting we do not necessarily trust a (single) peer. The Chaincode Enclave uses the Ledger Enclave as its tamperproof source of information on the blockchain world state. The Ledger Enclave and a Chaincode Enclave interact by through a secure channel that is initially established when the Chaincode Enclave is deployed. There is one Ledger Enclave per Channel on any given Peer. When a peer receives a new block from the ordering service, it forwards it to the Ledger Enclave. The Ledger Enclave is part of the FPC runtime.
+
+**FPC Registry:** Also referred to as the Enclave Registry Chaincode (ERCC), this is a component which maintains a list of all Chaincode Enclaves deployed on the peers in a channel.
+The registry associates with each enclave their identity, associated public keys and an attestation linking them. Additionally, the registry manages chaincode specific keys, including a chaincode public encryption key, and facilitates corresponding key-management among authorized Chaincode Enclaves. Lastly, the registry also records information required to bootstrap the validation of attestation. All of this information is committed firmly on the ledger. This enables any Peer in the Channel (even those without SGX) to inspect the Attestation results before taking actions such as connecting to that chaincode or committing transactions produced by an FPC chaincode. Moreover, clients can query the registry to retrieve the chaincode public encryption keys of a particular FPC chaincode so they can send privately transaction proposals for endorsement.
+
+**FPC Validator:** This validation complements the Peer validation logic by validating transactions produced by an FPC Chaincode.
+It does this by checking that each transaction has been cryptographically signed (endorsed) by a Chaincode Enclave that is listed in the FPC Registry and is authorized for that chaincode.
+
+**FPC Client SDK extension:** This extends the Fabric Client SDKs with extra functionality that allows users to write end-to-end secure FPC-based applications.
+In particular, the FPC Client SDK extension will provide these core functions: 1) FPC transaction proposal creation (including transparent encryption of arguments) 2) FPC transaction proposal response validation and decryption of the result.
+The encryption and decryption is performed by the Client SDK "under the covers" without requiring any special action by the users, i.e.,
+the users still use normal `invoke`/`query` functions to issue FPC transaction invocations.
+For the MVP, we intend to support the peer CLI only. Extended support for the NodeSDK (or Go SDK) will be future work.
```

Comment:
```suggestion
For the MVP, we intend to support the peer CLI only. Extended support for the NodeSDK (or Go SDK) will be future work.
```
While ok for the labs project to use the CLI, let's not make the mistake of adding this to the current peer "CLI".
Pick one SDK and implement it there.  I'd also say that you do not want to implement this inside the current SDKs ... meaning you should do something similar to how the higher level programming models work as a layer above the low level SDK code

Reply:
The FPC client-side request and response handling can be done transparently as pre/post-transformation of request/response in a proxy pattern way. Our original plan was to create two corresponding go-functions for this transformation and wrap them in a small utility which can be used in our peer wrapper to transparently do the processing for the `invoke` and `query` commands. As the cli is currently used for CI and is convenient for first walk-thru/tutorials, we will still do that.

However, we hear you and also added support for the Go SDK to the MVP plans. At a minimum level, we would just expose and document as separate library the two functions mentioned above. However, we will also investigate whether there is a straightforward way to implement a proxy on top of the normal sdk which will retain, with the exception of selecting a different package name or alike, the existing fabric client programming interface (it seems the mock implementation does something similar).